File: app/src/main/java/com/proxy/shadowsocksr/impl/TCPEncryptor.java
package com.proxy.shadowsocksr.impl;

import com.proxy.shadowsocksr.impl.crypto.CryptoManager;
import com.proxy.shadowsocksr.impl.crypto.crypto.AbsCrypto;

import java.util.Arrays;

public final class TCPEncryptor
{
    private final byte[] eIV;
    private boolean ivSent = false;
    private boolean ivNotRecv = true;

    private final AbsCrypto crypto;

    public TCPEncryptor(String pwd, String cryptMethod)
    {
        int[] cryptMethodInfo = CryptoManager.Companion.getCipherInfo(cryptMethod);
        eIV = Utils.Companion.srandomBytes(cryptMethodInfo[1]);
        crypto = CryptoManager.Companion.getMatchCrypto(cryptMethod, pwd);
    }

    public byte[] encrypt(byte[] buf)
    {
        if (buf.length == 0)
        {
            return buf;
        }
        if (ivSent)
        {
            return crypto.encrypt(buf);
        }
        ivSent = true;
        crypto.updateEncryptIV(eIV);
        buf = crypto.encrypt(buf);
        byte[] toSend = new byte[eIV.length + buf.length];
        System.arraycopy(eIV, 0, toSend, 0, eIV.length);
        System.arraycopy(buf, 0, toSend, eIV.length, buf.length);
        return toSend;
    }

    public byte[] decrypt(byte[] buf)
    {
        if (buf.length == 0)
        {
            return buf;
        }
        if (ivNotRecv)
        {
            if (buf.length < eIV.length + 1)
            {
                return new byte[0];
            }
            byte[] div = Arrays.copyOfRange(buf, 0, eIV.length);
            buf = Arrays.copyOfRange(buf, eIV.length, buf.length);
            crypto.updateDecryptIV(div);
            ivNotRecv = false;
        }
        return crypto.decrypt(buf);
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/impl/TCPEncryptor.kt
package com.proxy.shadowsocksr.impl

import com.proxy.shadowsocksr.impl.crypto.CryptoManager
import com.proxy.shadowsocksr.impl.crypto.crypto.AbsCrypto

import java.util.Arrays

class TCPEncryptor(pwd: String, cryptMethod: String)
{
    private val eIV: ByteArray
    private var ivSent = false
    private var ivNotRecv = true

    private val crypto: AbsCrypto

    val ivLen: Int get() = eIV.size

    init
    {
        val cryptMethodInfo = CryptoManager.getCipherInfo(cryptMethod)
        eIV = ImplUtils.srandomBytes(cryptMethodInfo[1])
        crypto = CryptoManager.getMatchCrypto(cryptMethod, pwd)
    }

    fun encrypt(buf: ByteArray): ByteArray
    {
        var bf = buf
        if (bf.size == 0)
        {
            return bf
        }
        if (ivSent)
        {
            return crypto.encrypt(bf)
        }
        ivSent = true
        crypto.updateEncryptIV(eIV)
        bf = crypto.encrypt(bf)
        val toSend = ByteArray(eIV.size + bf.size)
        System.arraycopy(eIV, 0, toSend, 0, eIV.size)
        System.arraycopy(bf, 0, toSend, eIV.size, bf.size)
        return toSend
    }

    fun decrypt(buf: ByteArray): ByteArray
    {
        var bf = buf
        if (bf.size == 0)
        {
            return bf
        }
        if (ivNotRecv)
        {
            if (bf.size < eIV.size + 1)
            {
                return ByteArray(0)
            }
            val div = Arrays.copyOfRange(bf, 0, eIV.size)
            bf = Arrays.copyOfRange(bf, eIV.size, bf.size)
            crypto.updateDecryptIV(div)
            ivNotRecv = false
        }
        return crypto.decrypt(bf)
    }
}
